pyproject.toml

```diff
@@ -46,6 +46,8 @@ markdown-it-py = { extras = ["plugins", "linkify"], version = ">=2.1.0" }
 #rich = {path="../rich", develop=true}
 importlib-metadata = ">=4.11.3"
 typing-extensions = "^4.4.0"
+tree-sitter = "^0.20.1"
+tree_sitter_languages = {version = ">=1.7.0", python = "^3.8"}
```

Comment:
While I'm not entitled to judge; this significantly complicates the packaging and distribution of Textual because is no longer pure Python.  Not only the TreeSitter Python bindings contain CPython extensions; but each language parser needs to be compiled.

So now, every time a new language is added to the text area, anyone repackaging or redistributing a Textual application will have to do additional work.  At least I know I will have to.

This is not really a complain and I don't want to stop progress.  However, I would like to call some attention to the fact that adding dependencies is never for free.

I do have a question, thoguh.  Why couldn't this widget have been implemented as a separate package or a Textual extension/plugin?

Reply:
tree-sitter and tree-sitter-languages have binary wheels available. You don't need to compile anything.

Reply:
For the average user yes but Isaac is talking about redistribution for example like Debian, Conda or Homebrew where everything is built from source

Reply:
It could still be factored out in to another package. It is something we considered. But then that is an extra hoop the end-user has to go through.

If Debian, Conda etc. are going to support the TextArea does it really matter if its part of the Textual core, or a separate package?

Reply:
In my case, I have a specially built CPython distribution for internal use.  One of the peculiarities is that, while CPython extensions are built as dynamically loadable libraries (`*.so`), any other external library is statically linked (e.g. Sqlite3, zlib, GNU dbm, OpenSSL, ...).  Even packages like lxml are built in a special way.

I will probably figure out how to build _tree-sitter_ in a way that it's compatible with my distribution.  After all, we love Textual because it's **awesome**.  My point is a bit more philosophical: depending on a 3rd party package is not all dandy, there are trade-offs (like introducing source build/packaging/distribution complexity in this case).

All I'm hoping is that these decisions are not made lightly.

Reply:
Glad you like Textual.

I do want to keep the dependancies lean in general. There are surprisingly few (non-dev) dependencies in Textual. Syntax highlighting is one of the areas where there just doesn't exist a pure-Python solution with adequate performance.

Reply:
Does this mean that TreeSitter will be favored and Pygments will be relegated to fall-back?  Will this impact rich as well?

Reply:
> All I'm hoping is that these decisions are not made lightly.

As someone who was in the same room when these discussions (plural) took place: I can assure that they're not.

Reply:
> Does this mean that TreeSitter will be favored and Pygments will be relegated to fall-back? Will this impact rich as well?

No, they are quite different APIs and not readily interchangeable. No need to change anything in Rich, since it doesn't need the incremental editing ability.